### ISSUE TITLE
NRE in IO.Ably.Transport.MsWebSocketTransport.ConnectAndStartListening and SemaphoreSlim.Wait

### ISSUE THREAD

**Issue by softsan** [ably/ably-dotnet#334]
We are getting lots of crash in our hockey app with following exception (on android) 
we are using ably 1.1.2

**Crashes per Day**
2290
Last 24 hours
6548
Last 7 days
6548
Last 30 days

```
MsWebSocketTransport.ConnectAndStartListening() - 2290 times in 24 hours
android.runtime.JavaProxyThrowable: System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. (Object reference not set to an instance of an object) ---> System.NullReferenceException: Object reference not set to an instance of an object
2a7c3906b6e14e42ae7028a3b0bcd816

Xamarin caused by: android.runtime.JavaProxyThrowable: System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. (Object reference not set to an instance of an object) ---> System.NullReferenceException: Object reference not set to an instance of an object
  at IO.Ably.Transport.MsWebSocketTransport.ConnectAndStartListening () [0x00189] in <2a7c3906b6e14e42ae7028a3b0bcd816>:0 
   --- End of inner exception stack trace ---
---> (Inner Exception #0) System.NullReferenceException: Object reference not set to an instance of an object
  at IO.Ably.Transport.MsWebSocketTransport.ConnectAndStartListening () [0x00189] in <2a7c3906b6e14e42ae7028a3b0bcd816>:0 <---
```

and this:
**Crashes per Day**
1251
Last 24 hours
3180
Last 7 days
3180
Last 30 days
 

```
SemaphoreSlim.Wait(System.Int32 millisecondsTimeout, System.Threading.CancellationToken cancellationToken)
android.runtime.JavaProxyThrowable: System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. (Value cannot be null.
cfcec5514a31434aae323caf1152dcac

Xamarin caused by: android.runtime.JavaProxyThrowable: System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. (Value cannot be null.
Parameter name: obj) ---> System.ArgumentNullException: Value cannot be null.
Parameter name: obj
  at (wrapper managed-to-native) System.Threading.Monitor.Exit(object)
  at System.Threading.SemaphoreSlim.Wait (System.Int32 millisecondsTimeout, System.Threading.CancellationToken cancellationToken) [0x0014a] in <cfcec5514a31434aae323caf1152dcac>:0 
  at System.Collections.Concurrent.BlockingCollection`1[T].TryTakeWithNoTimeValidation (T& item, System.Int32 millisecondsTimeout, System.Threading.CancellationToken cancellationToken, System.Threading.CancellationTokenSource combinedTokenSource) [0x00067] in <d07f211f06db4042a350980800838329>:0 
  at System.Collections.Concurrent.BlockingCollection`1+<GetConsumingEnumerable>d__68[T].MoveNext () [0x0004d] in <d07f211f06db4042a350980800838329>:0 
  at IO.Ably.Transport.MsWebSocketConnection+<>c__DisplayClass22_0.<StartSenderQueueConsumer>b__0 () [0x000e5] in <2a7c3906b6e14e42ae7028a3b0bcd816>:0 
   --- End of inner exception stack trace ---
---> (Inner Exception #0) System.ArgumentNullException: Value cannot be null.
Parameter name: obj
  at (wrapper managed-to-native) System.Threading.Monitor.Exit(object)
  at System.Threading.SemaphoreSlim.Wait (System.Int32 millisecondsTimeout, System.Threading.CancellationToken cancellationToken) [0x0014a] in <cfcec5514a31434aae323caf1152dcac>:0 
  at System.Collections.Concurrent.BlockingCollection`1[T].TryTakeWithNoTimeValidation (T& item, System.Int32 millisecondsTimeout, System.Threading.CancellationToken cancellationToken, System.Threading.CancellationTokenSource combinedTokenSource) [0x00067] in <d07f211f06db4042a350980800838329>:0 
  at System.Collections.Concurrent.BlockingCollection`1+<GetConsumingEnumerable>d__68[T].MoveNext () [0x0004d] in <d07f211f06db4042a350980800838329>:0 
  at IO.Ably.Transport.MsWebSocketConnection+<>c__DisplayClass22_0.<StartSenderQueueConsumer>b__0 () [0x000e5] in <2a7c3906b6e14e42ae7028a3b0bcd816>:0 <---
```

**Comment by caseyniemann** [comment]
Confirmed Ably 1.1 issue.  @withakay @mattheworiordan this seems to be similar to some of your other issues confirmed as bug.

**Comment by mattheworiordan** [comment]
@withakay this issue now has more details in support ticket 98229, which may help debug the problem.

@caseyniemann which issues are you referring to?  Unless you have a specific issue to look at, I am not sure it's going to be a good use of Jack's time to trawl through random issues to see if they may be related.  Which ones do you think are related and why?

**Comment by mattheworiordan** [comment]
@caseyniemann I believe you guys have been using 1.1.* for some time.  Is this issue localised to 1.1.2, or have you tested it and reproduced / confirmed the issue exists in previous versions too?

**Comment by lovetofail** [comment]
I believe this is the same issue as on screenshot. Happens for me after calling ablyRealtime.Close()
![2019-05-16_2252](https://user-images.githubusercontent.com/4531542/57882917-8a42b600-782d-11e9-90ad-1ce300c220eb.png)

**Comment by withakay** [comment]
@lovetofail thanks for your comment, very helpful.

I think have worked out the cause of the of the issue you have highlighted.
I am calling `Dispose` on `_sendQueue` whilst it is potentially being accessed by another thread :(
In actual fact we don't know when some other consuming thread might be finished with it _`sendQueue`, so calling `Dispose` is incorrect in this instance (This is somewhat counter intuitive Despite warnings to the contrary from the tooling) and just calling `CompleteAdding` is all that is required.

The offending code is here

https://github.com/ably/ably-dotnet/blob/91cad80a9748c184b1e25dc99ecbffd61ad9e898/src/IO.Ably.Shared/Transport/MsWebSocketConnection.cs#L214

I should be able to get a fix out tomorrow, could you possibly confirm that this resolves the issue at your end?

**Comment by lovetofail** [comment]
> could you possibly confirm that this resolves the issue

Seems to resolve it. Not getting the exception anymore.

**Comment by withakay** [comment]
@lovetofail please see #335 for a proposed fix. 
It works for me and tests are green so I don't think we have any regressions.
If you are ably to test that would be appreciated.

**Comment by lovetofail** [comment]
Seems to work fine. I'm getting ObjectDisposing exception. Please, let me know if it's ok. This happens on the second client.Close(). in a loop.
![2019-05-17_1725](https://user-images.githubusercontent.com/4531542/57934916-4a311100-78c9-11e9-8c78-ccff9c3ae33f.png)

**Comment by withakay** [comment]
@lovetofail Functionally that is fine, the code should work as expected. But, it is a bit ugly and could be dealt with better, so I will look into that.

**Comment by softsan** [comment]
@withakay @lovetofail @NVentimiglia @caseyniemann   - we upgraged to 1.1.3 of ably but still getting this exception: (It occurs two times)

```
MsWebSocketTransport.ConnectAndStartListening()
android.runtime.JavaProxyThrowable: System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. (Object reference not set to an instance of an object) ---> System.NullReferenceException: Object reference not set to an instance of an object
8d2322cefa174be1a27ff1b1555dedb6

```
```
Xamarin caused by: android.runtime.JavaProxyThrowable: System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. (Object reference not set to an instance of an object) ---> System.NullReferenceException: Object reference not set to an instance of an object
  at IO.Ably.Transport.MsWebSocketTransport.ConnectAndStartListening () [0x00189] in <8d2322cefa174be1a27ff1b1555dedb6>:0 
   --- End of inner exception stack trace ---
---> (Inner Exception #0) System.NullReferenceException: Object reference not set to an instance of an object
  at IO.Ably.Transport.MsWebSocketTransport.ConnectAndStartListening () [0x00189] in <8d2322cefa174be1a27ff1b1555dedb6>:0 <---
```

**Comment by withakay** [comment]
@softsan are you only seeing errors on Android? Are you deploying to iOS too, or just Android?

**Comment by withakay** [comment]
There is potential for `Logger` to be null in `MsWebSocketConnection` after `MsWebSocketTransport` has been disposed as it is being nulled in the `Dispose` method.

I have now observed this locally using shorter < 1000ms connect/disconnect cycles.
I have a branch that addresses this here https://github.com/ably/ably-dotnet/pull/338

It is the weekend now but I will try get a fresh release together with these changes ASAP.

Any further comments are welcome.

**Comment by softsan** [comment]
Yes on android no on iOS. @withakay

**Comment by withakay** [comment]
@softsan Just so I am clear, you have an iOS App and Android App, but are only seeing errors on Android, correct?

**Comment by softsan** [comment]
Yes @withakay we have iOS and android. But the above errors are occurring on Android.

**Comment by NVentimiglia** [comment]
@lovetofail @withakay 

@lovetofail is 100% correct. I think we may be using the IEnumerable incorrectly. Once we Await, we introduce a race condition where the collection can change. Moreover, if we signal out of this routine, (such as disconnect), we may still be continuing on that iteration (maybe not, but maybe, depends how we enter into this thread). 

I suggest we extract and cache the contents of the IEnumerable in a thread-safe way here; For instance, use a recursive algorithm to deque pending messages.

**Comment by caseyniemann** [comment]
@mattheworiordan see above.  Any options to simplify and resolve this weekend

**Comment by withakay** [comment]
@NVentimiglia 

> @lovetofail is 100% correct.

Correct about what, can you please be more specific?

> Maybe extract and cache the contents of the IEnumerable in a thread-safe way here

I assume you are referring to the the `sendQueue` here?
As per my previous response in PR #335, `BlockingCollection` is designed for this, the collection being modified is not the issue.
Please see https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.blockingcollection-1?view=netframework-4.8

I do agree that there are race conditions (mea culpa!), but I believe these are due to issues with the disposal of objects rather than the way we are managing the queue.

I have another PR inflight #338 that I am going to merge and release now.

**Comment by mattheworiordan** [comment]
> see above. Any options to simplify and resolve this weekend

@caseyniemann as you can see, we are working on this, on the weekend.  I am afraid you either rollback if it's urgent, or wait for a fix to come out.  There is nothing more we can do, especially given you don't have a support contract with Ably.

**Comment by withakay** [comment]
1.1.4 has been released to nuget.

**Comment by mattheworiordan** [comment]
@caseyniemann @softsan what's the latest on this bug and the 1.1.4 release?

**Comment by mattheworiordan** [comment]
@caseyniemann closing this issue.  Please drop me a line if this issue is still affecting you.
